pkg/rpcclient/unwrap/unwrap.go

```diff
@@ -180,6 +180,53 @@ func SessionIterator(r *result.Invoke, err error) (uuid.UUID, result.Iterator, e
 	return r.Session, iter, nil
 }
 
+// ArrayAndSessionIterator expects correct execution (HALT state) with one or two stack
+// items returned. If there is 1 item, it must be an array. If there is a second item,
+// it must be an iterator. This is exactly the result of smartcontract.CreateCallAndPrefetchIteratorScript.
+// NeoGo server SessionsEnabled parameter must be set to true.
+func ArrayAndSessionIterator(r *result.Invoke, err error) ([]stackitem.Item, uuid.UUID, result.Iterator, error) {
+	if err := checkResOK(r, err); err != nil {
+		return nil, uuid.UUID{}, result.Iterator{}, err
+	}
+	if r.State != vmstate.Halt.String() {
+		return nil, uuid.UUID{}, result.Iterator{}, fmt.Errorf("invocation failed: %s", r.FaultException)
+	}
+	if len(r.Stack) == 0 {
+		return nil, uuid.UUID{}, result.Iterator{}, errors.New("result stack is empty")
+	}
+	if len(r.Stack) != 1 && len(r.Stack) != 2 {
+		return nil, uuid.UUID{}, result.Iterator{}, fmt.Errorf("expected 1 or 2 result items, got %d", len(r.Stack))
+	}
+
+	// Unwrap array.
+	itm := r.Stack[0]
+	if t := itm.Type(); t != stackitem.ArrayT {
+		return nil, uuid.UUID{}, result.Iterator{}, fmt.Errorf("expected Array, got %s", t)
+	}
+	arr, ok := itm.Value().([]stackitem.Item)
+	if !ok {
+		return nil, uuid.UUID{}, result.Iterator{}, errors.New("the item has Array type, but not is not an array")
+	}
+
+	// Check whether iterator exists and unwrap it.
+	if len(r.Stack) == 1 {
+		return nil, uuid.UUID{}, result.Iterator{Values: arr}, nil
+	}
+
+	itm = r.Stack[1]
+	if t := itm.Type(); t != stackitem.InteropT {
+		return nil, uuid.UUID{}, result.Iterator{}, fmt.Errorf("expected InteropInterface, got %s", t)
+	}
+	iter, ok := itm.Value().(result.Iterator)
+	if !ok {
+		return nil, uuid.UUID{}, result.Iterator{}, errors.New("the item is InteropInterface, but not an Iterator")
+	}
+	if (r.Session == uuid.UUID{}) && iter.ID != nil {
+		return nil, uuid.UUID{}, result.Iterator{}, ErrNoSessionID
+	}
```

Comment:
AFAIU, the case when sessions are disabled on the server side is a no-op for this method. If that's true, then here we need to require `r.Session` not to be empty.

This part of code was added to `unwrap.SessionIterator` because it may happen that sessions are disabled on the server side, and in this case we can still get iterator items (up to max count) unwrapped and placed instead of the corresponding iterator stackitem.

But `ArrayAndSessionIterator` needs sessions to be enabled anyway, thus `r.Session == uuid.UUID{}` case is always no-op for it. Right?

Reply:
Take a look at the https://github.com/nspcc-dev/neo-go/blob/de98b39a9502d64ab797a9191891955d7b8df21e/pkg/services/rpcsrv/server.go#L2492

Reply:
Rewrote the check